### DISCUSSION TITLE
Hooks for actions that update local and remote commit status

### DISCUSSION THREAD

**Discussion by Alexander-Miller** [magit/magit#4699] [Ideas]
I am currently in the process of building a mode for treemacs that shows you how many commits your local repository is ahead and/or behind its remote, a feature you probably know from other editors:

![diff](https://user-images.githubusercontent.com/5951157/171044315-d9c6f5dd-e264-438e-82e0-e0d06d11b66f.png)

Getting that information is easy, I just need to look at `git status -sb`. Knowing when to get it the hard part. Doing it on every treemacs refresh is not automatic and also wasteful. Same goes for `magit-post-refresh-hook`. The best solution is to only make updates when really necessary. Treemacs already has this kind of integration with magit, updating itself whenever files are staged or committed, so what I am looking for are further hooks for the kind of actions that would change the ahead/behind counter. From the top of my head that would mean post-event hooks for
 - pulling
 - fetching
 - merging
 - committing
 - rebasing
 - switching branches

and anything else I may have forgotten.

Performance won't be an issue, all the work of waiting for git and parsing its output happens in an external python process, the api to annotate treemacs' nodes is optimized to not get in the way. 

WDYT?

**Comment by tarsius** [comment]
I am aware that I have to teach Magit to "only update the things that actually need to be updated" eventually. I don't think that doing that will become easier if I already have added many hooks that let third-party packages do that for their more narrow use-cases, which I then have to preserve even if I take a different approach for the sake of Magit itself.

Some of the mentioned commands already provide hooks that you could use. For the others you would have to add advises.

> Doing it on every treemacs refresh is not automatic and also wasteful. Same goes for magit-post-refresh-hook. 

Compared to what magit is doing, this doesn't seem so bad. I would recommend you just live with the waste for now or require the user to explicitly refresh when they think that is necessary.

**Comment by Alexander-Miller** [comment]
> I am aware that I have to teach Magit to "only update the things that actually need to be updated" eventually.

That's not the goal for me. I'm looking to synchronize treemacs not with magit, but git itself. So the hooks I'm looking for are not magit-event-hooks, but rather git-process-finished hooks, especially for the difficult to catch async stuff like fetch and pull.

Official hooks would be nice, but I suppose advice is also an option. I'll try my luck with`magit-process-sentinel`, it looks like it has all the info I need. Just one more thing: is there a better way to determine which command was run that going through `(process-command process)`? It looks like there's a lot of standard arguments included in that list.

**Reply by tarsius** [reply]
> Official hooks would be nice, but I suppose advice is also an option. I'll try my luck withmagit-process-sentinel, it looks like it has all the info I need. 

And `magit-call-git`.

> Just one more thing: is there a better way to determine which command was run that going through (process-command process)? It looks like there's a lot of standard arguments included in that list.

You can use the same logic as in `magit-process-set-mode-line`.
